_drafts/id-generation.md

```diff
@@ -0,0 +1,104 @@
+
+Over the course of this article, we'll walk you through how we arrived at a solution for generating IDs in a distributed fashion using the Redis instances we already had deployed and in a format that allows _k_-ordering. We'll share with you the code for doing this so you can use this ID generation in your own projects!
+
+## What is an ID?
+
+An ID can be many things, but surely its most important definition is that of "uniqueness". Of all the properties, this is the only one that truly matters. If you label a piece of data with an ID, you want to be sure you don't later have ambiguities when looking up that data again by the ID you had.
+
+IDs may have some other useful properties, however:
+
+* _Ordered:_ They may be ordered to allow comparison of items stamped with IDs.
+
+* _Meaningful:_ They may embed certain information about the point at which the ID was generated, for example the time (however inexact).
+
+* _Distributed:_ They may be generated by a machine without consulting any other machine (ie. they must be truly distributed).
+
+* _Compact:_ They may be represented in a specific format for space or performance reasons. Normally, you'd want to represent this in as compact a format as possible but you often trade-off longevity to do this ([32-bit UNIX timestamps are a great example of this](https://en.wikipedia.org/wiki/Year_2038_problem)).
+
+These all sound like useful properties to have in an ID, but existing methods don't really work to preserve these properties in a distributed environment. Traditionally, in software, we have generated IDs using methods that do not scale. Either we let an application like a database generate IDs for us, in which case we (generally) have a single point-of-failure, or we use standards like the UUID which [do not guarantee uniqueness in a distributed environment without significant efforts at coordination](http://en.wikipedia.org/wiki/Universally_unique_identifier#Random_UUID_probability_of_duplicates).
+
+As a turns out, primarily, the reasons for doing this are deeply rooted in our inability to keep time across a distributed environment.
+
+## Time is hard
+
+Keeping time is actually a really hard problem to solve. At the core of the issue is the fact that computers suffer from "clock drift", where the clock is constantly skewing forwards or backwards away from the actual time. Perhaps even worse, [every computer skews at a different rate](https://en.wikipedia.org/wiki/Clock_skew).
+
+Thus, the concept of a "time oracle" is born. Time oracles dedicated servers normally equipped with an atomic clock or GPS device whose sole purpose is for keeping time in as precise a manner as possible. Time oracle servers are used frequently in environments where super accurate time is important (such as trading platforms and financial applications), but they are prohibitively expensive and not a viable solution for those unable to manage their own physical server.
```

Comment:
More than "those unable to manage their own physical server" I would probably write "those we don't manage their own physical server".